### PR TITLE
Scrapper implementated for multiple sites

### PR DESCRIPTION
**Pull request by samarthdd**
Deliverables:
Scraping of multiple sites and configuration of the same from .env.
The following sites are covered:
https://virusshare.com/
https://github.com/ytisf/theZoo
http://www.tekdefense.com/downloads/malware-samples/
https://das-malwerk.herokuapp.com/

2. Storage
Storage using minio adapter.
Malware Zip files are uploaded to minio directly.

### CODE REVIEW

**Comment by DinisCruz** [comment]
Please fix conflict

File: gw_crawler/malicious_file_crawler/src/pipelines.py [hunk 1]
```diff
@@ -2,56 +2,67 @@
...
-        checksum = md5sum(buf)
-        buf.seek(0)
-        self.store.persist_file(path, buf, info)
-
-        file_path = zip_download_path + "/" + path
-        key = path.split("/")[-1].split(".")[0]
+        try:
+            logger.info(f'MaliciousFileCrawlerPipeline:file_downloaded:: Response status {response.status}')
+            if (response.status == 200):
+                if (info.spider.name == "virusshare"):
```
Comment (susilupwork) [inline gw_crawler/malicious_file_crawler/src/pipelines.py:35]: No hard code logic for virusshare. Generalize it.

File: gw_crawler/malicious_file_crawler/src/pipelines.py [hunk 2]
```diff
@@ -2,56 +2,67 @@
...
+            Store object in minio
+        """
+        try:
             client = MinioClient.get_client()
             if not client.bucket_exists(bucket_name):
                 client.create_bucket(bucket_name)
-            client.upload_file(bucket_name, path.split("/")[-1], zipped_file)
 
-        return checksum
+            if not client.bucket_exists(bucket_name):
```
Comment (susilupwork) [inline gw_crawler/malicious_file_crawler/src/pipelines.py:63]: This should be minIO service code. It should be there not at the calling end.

File: gw_crawler/malicious_file_crawler/src/spiders/malshare_crawler.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            malhare url http://www.malshare.com
+            api url http://www.malshare.com/api.php
+            Get the malware url using hashes and api and send it to storage
+    """
+
+    name = 'malshare'
+    # Allow duplicate url request (we will be crawling "page 1" twice)
+    # custom_settings will only apply these settings in this spider
+    custom_settings = {
+        'DUPEFILTER_CLASS': 'scrapy.dupefilters.BaseDupeFilter',
```
Comment (susilupwork) [inline gw_crawler/malicious_file_crawler/src/spiders/malshare_crawler.py:22]: Should go to BaseScrapper.

File: gw_crawler/malicious_file_crawler/src/spiders/malshare_crawler.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        self.cfg = config
+        self.API_KEY = self.cfg.get('api_key')
+        self.API_URL = self.cfg.get('api_base_url')
+        self.base_url=self.cfg.get('base_url')
+        self.url=self.cfg.get('api_url')
+
+
+    def start_requests(self):
+        logger.info(f'Site url : {self.base_url}')
+        yield scrapy.Request(url=self.base_url, callback=self.parser)
```
Comment (susilupwork) [inline gw_crawler/malicious_file_crawler/src/spiders/malshare_crawler.py:37]: Add Exception handling and log if exception occurs

File: gw_crawler/malicious_file_crawler/storage/config/config.ini [hunk 5]
```diff
@@ -1,12 +1,11 @@
 [MINIO]
-HOSTNAME=minio:9000
-AWS_ACCESS_KEY_ID=minio1
-AWS_SECRET_ACCESS_KEY=minio1@123
+HOSTNAME=Please provide Hostname
```
Comment (susilupwork) [inline gw_crawler/malicious_file_crawler/storage/config/config.ini:2]: No key in config file. Only in .env file as sample.

File: gw_crawler/malicious_file_crawler/storage/src/minio_adapter.py [hunk 6]
```diff
@@ -1,16 +1,24 @@
...
         # Initialize minioClient with an endpoint and access/secret keys.
 
         super().__init__(*args, **kwargs)
-        self._client = Minio(endpoint=config.get('HOSTNAME'),
-                             access_key=config.get('AWS_ACCESS_KEY_ID'),
-                             secret_key=config.get('AWS_SECRET_ACCESS_KEY'),
-                             secure=False)
+        try:
+            self._client = Minio(endpoint=config.get('HOSTNAME'),
+                                 access_key=config.get('AWS_ACCESS_KEY_ID'),
```
Comment (susilupwork) [inline gw_crawler/malicious_file_crawler/storage/src/minio_adapter.py:16]: MiniIO is not going to store in AWS S3 directly.

**Comment by samarthdd** [comment]
> Please fix conflict



> Please fix conflict

Done.I have fixed conflict and commited the same

File: gw_crawler/malicious_file_crawler/src/utils/minio_client.py [hunk 7]
```diff
@@ -2,20 +2,28 @@
...
 class MinioClient:
+    """
+            It calls storage adapter to get minio client
+    """
+
     @staticmethod
     def get_client():
-        client = MinioService.get_storage_adapter()
-        return client
+        try:
```
Comment (susilupwork) [inline gw_crawler/malicious_file_crawler/src/utils/minio_client.py:23]: The MinIO client should call only the MinIO endpoint. The scrapper will only deal with MinIO end point, no direct calling.

**Review by susilupwork — CHANGES_REQUESTED** [review]
- All Key must come from environment variable. use config.py to load them. 
- Virushare the scrapping constraint should be in config. 1 request per 15 sec. Also there should global scrapping should be for all sites , say 10 sec for each request. We do not want anybody gets down due to us.
- Do not use site name as MinIObucketname. You can use original zip or similar. 
- The MinIO client should call only the MinIO endpoint. The scrapper will only deal with MinIO end point, no direct calling. 
- Many places still Exception handling  is not there, add and log if exception occurs.
- Go through GW coding guideline . Many violations noticed in constant name and class name etc.

File: gw_crawler/malicious_file_crawler/src/pipelines.py [hunk 8]
```diff
@@ -2,56 +2,67 @@
...
+
 BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
 
 
 class MaliciousFileCrawlerPipeline(FilesPipeline):
 
     def file_downloaded(self, response, request, info, unzip_path=None):
-        bucket_name = os.environ['bucket_name']
+        """
+            Downloads file
```
[the remark this answers is not in the document]
Comment (samarthdd) [reply gw_crawler/malicious_file_crawler/src/pipelines.py:25]: This is removed as you see the same in red highlight

File: gw_crawler/malicious_file_crawler/src/config/config.ini [hunk 9]
```diff
@@ -1,17 +1,43 @@
...
-name = tek_defence_spider
+name = tekdefence
 login_url = http://www.tekdefense.com/downloads/malware-samples/
 file_page_url = http://www.tekdefense.com/downloads/malware-samples/
 
+[VIRUSSHARE]
+name = virusshare
+virusshare_url = https://virusshare.com/apiv2/{}?apikey={}&hash={}
+virusshare_hash_url = https://virusshare.com/hashfiles/unpacked_hashes.md5
+vs_api_key=enter virus share api key here
```
[the remark this answers is not in the document]
Comment (samarthdd) [reply gw_crawler/malicious_file_crawler/src/config/config.ini]: Resolved in latest commit

File: gw_crawler/malicious_file_crawler/src/constants.py [hunk 10]
```diff
@@ -5,3 +5,5 @@
 zip_download_path = BASE_DIR + "/static/"
 base_unzip_path = BASE_DIR + "/static/unzip"
 zip_path = BASE_DIR + "/static/zip"
+
+download_path = BASE_DIR + "/static/"
```
[the remark this answers is not in the document]
Comment (samarthdd) [reply gw_crawler/malicious_file_crawler/src/constants.py]: Resolved in latest commit

File: gw_crawler/malicious_file_crawler/src/glasswall_crawler_runner.py [hunk 11]
```diff
@@ -7,7 +7,7 @@
 from scrapy.exceptions import CloseSpider
 from scrapy.crawler import CrawlerProcess
 from scrapy.utils.project import get_project_settings
-from malicious_file_crawler.src.utils.read_config import ConfigReader
+from src.utils.read_config import ConfigReader
 
 logging.basicConfig(filename="testdata.log",
```
[the remark this answers is not in the document]
Comment (samarthdd) [reply gw_crawler/malicious_file_crawler/src/glasswall_crawler_runner.py]: Resolved

File: gw_crawler/malicious_file_crawler/src/pipelines.py (hunk 1 above)
Comment (samarthdd) [reply gw_crawler/malicious_file_crawler/src/pipelines.py]: I have removed hard code logic and has implemented generalized logic

File: gw_crawler/malicious_file_crawler/src/pipelines.py (hunk 2 above)
Comment (samarthdd) [reply gw_crawler/malicious_file_crawler/src/pipelines.py]: Removed it and updated in latest commit

File: gw_crawler/malicious_file_crawler/src/spiders/malshare_crawler.py (hunk 4 above)
Comment (samarthdd) [reply gw_crawler/malicious_file_crawler/src/spiders/malshare_crawler.py]: Added exception handling for all methods

File: gw_crawler/malicious_file_crawler/storage/src/minio_adapter.py (hunk 6 above)
Comment (samarthdd) [reply gw_crawler/malicious_file_crawler/storage/src/minio_adapter.py]: Removed

File: gw_crawler/malicious_file_crawler/storage/config/config.ini (hunk 5 above)
Comment (samarthdd) [reply gw_crawler/malicious_file_crawler/storage/config/config.ini]: removed keys in config

File: storage/config/config.ini [hunk 12]
```diff
@@ -0,0 +1,11 @@
+[MINIO]
+HOSTNAME=Please provide Hostname
```
Comment (susilupwork) [inline storage/config/config.ini:2]: No KEY in config file, take it to to environment variable

**Review by susilupwork — APPROVED** [review]
Please do this changes after 29th demo 

- DO not hard code values, It should come from config file. 
- Do not use sleep, instead use DOWNLOAD_DELAY. The default delay should be there for every site.
- No KEY in config file, take it to to environment variable

File: storage/config/config.ini (hunk 12 above)
Comment (samarthdd) [reply storage/config/config.ini]: Updated

File: gw_crawler/malicious_file_crawler/src/spiders/virus_share.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+            raise error
+
+    def parser(self, response):
+        try:
+            logger.debug("download_files")
+
+            hashes = self.scrape_hashes()
+
+            for _hash in hashes:
+                sleep(15)
```
[the remark this answers is not in the document]
Comment (samarthdd) [reply gw_crawler/malicious_file_crawler/src/spiders/virus_share.py]: Updated with download_dealy

File: gw_crawler/malicious_file_crawler/src/spiders/dasmalwerk_crawler.py [hunk 14]
```diff
@@ -23,19 +25,30 @@ def __init__(self, config=None, data=None):
...
         # get download file link
-        html = html_xml.fromstring(response.text)
-        file_download_link_elements = html.xpath("//tr//td[2]/a/@href")
-
-        loader = ItemLoader(item=MaliciousFileCrawlerItem())
-
-        for link in file_download_link_elements:
-            loader.add_value('file_urls', link)
-            yield loader.load_item()
+        try:
```
[the remark this answers is not in the document]
Comment (samarthdd) [reply gw_crawler/malicious_file_crawler/src/spiders/dasmalwerk_crawler.py:44]: added info logging in scraper

**Review by dtollaku — APPROVED** [review]

**Pull request merged**